### PR TITLE
feat(qwen35): support d2t draft-vocabulary trim for MTP speculative decoding

### PR DESCRIPTION
**Pull request by Shreesh-Sree**
## Overview

This PR adds support for trimmed/compact draft vocabularies via the `d2t` (draft-to-target) mapping tensor in both `qwen35` and `qwen35moe` model architectures during MTP (Multi-Token Prediction) speculative decoding.

When loading an MTP draft model or sidecar trained with a reduced draft vocabulary (such as FastMTP / compact speculative heads), the LM head (`output.weight` or `layer.nextn.shared_head_head`) has dimensions `[n_embd, n_draft_vocab]` where `n_draft_vocab < n_vocab`. Previously, `llama_model_qwen35` and `llama_model_qwen35moe` unconditionally initialized output projections with `n_vocab`, causing loader dimension mismatches or runtime assertion failures.

With this PR:
1. When `d2t` is present, `output` and `layer.nextn.shared_head_head` are initialized with the draft vocabulary dimension `n_vocab_out = d2t->ne[0]`.
2. Fallback to `output = create_tensor(tn(LLM_TENSOR_TOKEN_EMBD), ...)` is prevented when `d2t` is present and an explicit draft head is provided.
3. In `graph_mtp`, draft candidate logits of shape `[n_draft_vocab, n_outputs]` are scattered onto a full-vocabulary `-inf` tensor of shape `[n_vocab_full, n_outputs]` using `ggml_set_rows`, consistent with existing `dflash.cpp` and `eagle3.cpp` implementations.
4. Validation assertions confirm that draft vocabulary dimensions match the active LM head.

---

### 1. Root Cause
In standard MTP, the next-n prediction head shares the full vocabulary (`n_vocab = 152064` for Qwen 3.5). However, high-throughput speculative decoding optimizations often prune the draft vocabulary to the most frequent tokens (e.g. 32,768 tokens) mapped by a `d2t` tensor.
- Without `d2t` handling, `load_arch_tensors` expected `[n_embd, 152064]`, causing GGUF tensor loading failure (`tensor 'blk.0.nextn.shared_head_head.weight' size mismatch`).
- If fallback occurred, `tok_embd` was assigned to `output`, clobbering the compact draft head with target embedding weights.

### 2. Implementation
- **Detection:** `load_arch_tensors` inspects `ml.get_tensor_meta("d2t")` when `mtp_only` is true.
- **Tensors:** `output` and `layer.nextn.shared_head_head` allocate `n_vocab_out` rows.
- **Logits Projection:** In `graph_mtp`, after `cur = build_lora_mm(head_w, cur, head_s)`:
  ```cpp
  if (model.d2t) {
      const int64_t n_draft_vocab = cur->ne[0];
      const int64_t n_outputs     = cur->ne[1];
      const int64_t n_vocab_full  = (int64_t) model.vocab.n_tokens();

      GGML_ASSERT(model.d2t->ne[0] == n_draft_vocab);

      ggml_tensor * logits = ggml_fill(ctx0,
              ggml_new_tensor_3d(ctx0, GGML_TYPE_F32, 1, n_vocab_full, n_outputs), -INFINITY);
      cur = ggml_set_rows(ctx0, logits,
              ggml_reshape_3d(ctx0, cur,       1,             n_draft_vocab, n_outputs),
              ggml_reshape_3d(ctx0, model.d2t, n_draft_vocab, 1,             1));
      cur = ggml_reshape_2d(ctx0, cur, n_vocab_full, n_outputs);
      cb(cur, "result_output_d2t", -1);
  }
  ```

---

## Benchmarks & Verification

Tested on dual NVIDIA RTX PRO 6000 Ada / Blackwell GPUs running live inference:
- **Baseline (Autoregressive dense decoding):** ~44.5 tokens/sec (22.47 ms/token).
- **With FastMTP Draft-Vocab Trim:** **102.2 – 113.9 tokens/sec** (**8.78 – 9.84 ms/token**).
- **Speedup:** **~2.3x – 2.5x throughput improvement** with up to **82.89%** draft acceptance rate.
- **Compilation:** Clean build verified on GCC 13.3, Clang 18, and NVCC CUDA 12.0/13.3.

---

## Authors & Attribution
- **Sreesanth R** ([@Shreesh-Sree](https://github.com/Shreesh-Sree))
- **Allen Joseph** ([@unknownhackerworld](https://github.com/unknownhackerworld))

---

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES - AI assistance was used for drafting benchmarks, architectural diagrams, and documentation; all code changes were implemented, verified, and benchmarked on physical hardware.

### CODE REVIEW

**Comment by Shreesh-Sree** [comment]
Closes #29145

**Comment by Eithoo** [comment]
Same use case here - FastMTP-style MTP draft models for qwen35 where the output projection is over a reduced (32K) draft vocabulary with a d2t tensor. I had opened a duplicate PR (#29290, now closed); yours also covers qwen35moe and the shared_head_head path, so I'll defer to this one. I can confirm the d2t approach works in my setup (llama-server with --spec-type draft-mtp, 32K draft-vocab GGUF). Happy to help with testing.
